tests/core/jobrun_test.py

```diff
@@ -676,3 +679,255 @@ def test_get_action_runs(self):
+
+    def test_one_action_unknown(self, job_run):
+        foo = job_run.get_action_run('foo')
+        after_foo = job_run.get_action_run('after_foo')
+        bar = job_run.get_action_run('bar')
+
+        # bar action becomes unknown, job is RUNNING because foo is still running
+        job_run.start()
+        foo.action_command.started()
+        bar.action_command.started()
+        bar.action_command.exited(None)
+        assert job_run.state == actionrun.ActionRun.RUNNING
+
+        # after_foo still starts after its dependency succeeds
+        foo.action_command.exited(0)
+        assert after_foo.is_starting
+        after_foo.action_command.started()
+        assert job_run.state == actionrun.ActionRun.RUNNING
+
+        # UNKNOWN after running actions finish
+        after_foo.action_command.exited(0)
+        assert job_run.state == actionrun.ActionRun.UNKNOWN
+
+    def test_both_unknown_and_failed(self, job_run):
+        foo = job_run.get_action_run('foo')
+        after_foo = job_run.get_action_run('after_foo')
+        bar = job_run.get_action_run('bar')
+
+        # bar action becomes unknown, job is RUNNING because foo is still running
+        job_run.start()
+        foo.action_command.started()
+        bar.action_command.started()
+        bar.action_command.exited(None)
+        assert job_run.state == actionrun.ActionRun.RUNNING
+
+        # after_foo still starts after its dependency succeeds
+        foo.action_command.exited(0)
+        assert after_foo.is_starting
+        after_foo.action_command.started()
+        assert job_run.state == actionrun.ActionRun.RUNNING
+
+        # A different action fails
+        # Overall run is FAILED
+        after_foo.action_command.exited(1)
+        assert job_run.state == actionrun.ActionRun.FAILED
+
+    def test_required_action_fails(self, job_run):
+        foo = job_run.get_action_run('foo')
+        after_foo = job_run.get_action_run('after_foo')
+        bar = job_run.get_action_run('bar')
+
+        # An action (foo) required by another action fails
+        # Run is RUNNING while the other action, bar, is running
+        job_run.start()
+        foo.action_command.started()
+        bar.action_command.started()
+        foo.action_command.exited(1)
+        assert job_run.state == actionrun.ActionRun.RUNNING
+
+        # bar action succeeds
+        # after_foo cannot run because its required action failed
+        # So run is FAILED even though after_foo is waiting
+        bar.action_command.exited(0)
+        assert after_foo.is_waiting
+        assert job_run.state == actionrun.ActionRun.FAILED
+
+        # Pretend we reconfigured and after_foo doesn't depend on foo anymore
+        # Run should not be waiting
+        # Ideally it would still be FAILED, but for now it's UNKNOWN in this case
+        job_run.action_runs.action_graph.required_actions['after_foo'] = {}
+        assert job_run.state != actionrun.ActionRun.WAITING
```

Comment:
This line seems redundant given the next. A comment explaining that it should be waiting is enough.